### PR TITLE
Test Compression Algorithm Selection Refactor

### PR DESCRIPTION
**Pull request by Ornias1993**
This PR is based on my previous work in #9626 trying to find a more standardised way to select compression algorithms for the test suit.

TL:DR: 
- Removes duplicate algorithms from compression tests
- Refactors compression algorithm selection for tests to use one uniform format

Signed-off-by: Kjeld Schouten-Lebbing <kjeld@schouten-lebbing.nl>

### Motivation and Context
During my work on investigating the compression test suite for #8941 a few things became appearant:
1. There are a multitude of different ways/arrays/variables containing compression algorithms for tests
2. Already some algorithms seem to have skipped over certain tests because of this
3. Listing multiple (gzip) levels leads to duplicate and decreases the chance of other algorithms being choosen
4. The decreased chance for different algorithms increases a lot when other "level based" algorithms get added
5. Listing aliasses of algorithms leads to duplicate tests
6. Somehow someone had the idea to use a function to simply pass array values to a for loop.

### Description

This PR refactors it all.

- The only compression algorithm list left is the one inside `properties.shlib`
- It removes all seperate levels from the compression algorithms list for tests, except one (the default)
- It removes the `on` alias for the selection of tests, in favor of `lz4` to prevent mistakes switching default algorithm
- It removes random algorithm selection, if the new array is just 1 or 2 longer than the previous amount of random draws
- It removes the function onder `libtest.shlib` that was used to pass compresion algorithms (now directly uses the array in `properties.shlib` instead)
- As a consequence, all tests now use the right set of algorithms
- Also cleans the copyright notices a bit, layout was not always the best

**One list to use them and in compression store them**

### Is * taken into account aka FAQ?
Q: Can't test failure of different levels slip past now? 
_A: Mostly levels are similair enough to the default to be found out just by using the default. If that's not the case a seperate test testing the different levels one time is more efficient than doing so for all tests._

Q: Can't test failure of aliasses slip past now?
_A: Aliasses already have their own tests script or (in case of `on`) still get used a few times directly. So failures shouldn't slip past._

Q: Don't we need more randomised selection again when more compression algorithms get added?
_A: If the amount of algorithms in the list increases byond 8 or 9, 1 or 2 extra random draws might be needed. Considering the current number and the (slow) speed of new algorithms being added, this will suffice for quite a while_

Q: Doesn't adding the same complete selection of algorithms increase test runtime?
_A: In scenario's where some compression algorithms where left out and no randomisation is used, it might take a little longer. However the same subset of algorithms should already have been used and, by decreasing the amount of gzip-based algorithms in the selection, in general less performance intensive tests are being used. Net it will stay about the same or a little faster, but in general the test will be more repeatable and comparable.

### How Has This Been Tested?
I manually checked if all tests actually used the algorithms referenced and they do.

### Types of changes
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Performance enhancement (non-breaking change which improves efficiency)
- [X] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
- [X] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [X] I have updated the documentation accordingly.
- [X] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [X] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [X] All new and existing tests passed.
- [X] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by Ornias1993** [comment]
@c0d3z3r0 
This is just a quick draft, But I guess this is more like it...

**Comment by Ornias1993** [comment]
@behlendorf 
This is what I had in mind when closing #9626

**Comment by Ornias1993** [comment]
*edit to description*
Forgot to change the "how has this been tested"...

I manually checked all edited tests to make sure they actually got the compression algorithms referenced.

Also nice to note for review:
refactor uses one of the many different reference styles used by @dankimmel from https://github.com/zfsonlinux/zfs/commit/a7004725d0ad52fa005c29f68ce55273f11d95ac and uses that one to unify compression algorithm references (including other styles included)

So this reference style is not new, it just unifies different reference styles.

File: tests/zfs-tests/tests/functional/nopwrite/nopwrite_varying_compression.ksh [hunk 1]
```diff
@@ -52,7 +53,7 @@ dd if=/dev/urandom of=$TESTDIR/file bs=1024k count=$MEGS conv=notrunc \
     >/dev/null 2>&1 || log_fail "initial dd failed."
 
 # Verify nop_write for 4 random compression algorithms
```
Comment (c0d3z3r0) [inline tests/zfs-tests/tests/functional/nopwrite/nopwrite_varying_compression.ksh]: `for 4 random`

File: tests/zfs-tests/tests/functional/compression/compress_004_pos.ksh [hunk 2]
```diff
@@ -22,13 +22,12 @@
 
 #
 # Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
-# Use is subject to license terms.
-#
-
-#
 # Copyright (c) 2013, 2016 by Delphix. All rights reserved.
+# Copyright (c) 2019 Kjeld Schouten-Lebbing, All Rights Reserved.
```
Comment (c0d3z3r0) [inline tests/zfs-tests/tests/functional/compression/compress_004_pos.ksh]: `Lebbing, All` -> `Lebbing. All`

File: tests/zfs-tests/tests/functional/compression/compress_004_pos.ksh [hunk 3]
```diff
@@ -22,13 +22,12 @@
 
 #
 # Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
```
Comment (c0d3z3r0) [inline tests/zfs-tests/tests/functional/compression/compress_004_pos.ksh]: fix that duplicate space, please

File: tests/zfs-tests/tests/functional/compression/compress_003_pos.ksh [hunk 4]
```diff
@@ -21,14 +21,14 @@
 #
 
 #
-# Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
+# Copyright (c) 2007 Sun Microsystems, Inc.  All rights reserved.
```
Comment (c0d3z3r0) [inline tests/zfs-tests/tests/functional/compression/compress_003_pos.ksh]: double space

File: tests/zfs-tests/tests/functional/compression/compress_003_pos.ksh [hunk 5]
```diff
@@ -21,14 +21,14 @@
...
+# Copyright (c) 2013, 2016 by Delphix. All rights reserved.
+# Copyright (c) 2019 Kjeld Schouten-Lebbing, All Rights Reserved.
```
Comment (c0d3z3r0) [inline tests/zfs-tests/tests/functional/compression/compress_003_pos.ksh]: period instead of comma

File: tests/zfs-tests/tests/functional/cli_root/zfs_set/property_alias_001_pos.ksh [hunk 6]
```diff
@@ -22,13 +22,11 @@
 
 #
 # Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
```
Comment (c0d3z3r0) [inline tests/zfs-tests/tests/functional/cli_root/zfs_set/property_alias_001_pos.ksh]: another duplicated space

File: tests/zfs-tests/tests/functional/cli_root/zfs_set/property_alias_001_pos.ksh [hunk 7]
```diff
@@ -22,13 +22,11 @@
...
-# Use is subject to license terms.
-#
-
-#
 # Copyright (c) 2016, 2017 by Delphix. All rights reserved.
```
Comment (c0d3z3r0) [inline tests/zfs-tests/tests/functional/cli_root/zfs_set/property_alias_001_pos.ksh]: proudly presented ***by*** ;) you know what I mean :D

File: tests/zfs-tests/tests/functional/rsend/rsend_012_pos.ksh [hunk 8]
```diff
@@ -22,13 +22,12 @@
 
 #
 # Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
```
Comment (c0d3z3r0) [inline tests/zfs-tests/tests/functional/rsend/rsend_012_pos.ksh]: and another redundant space

File: tests/zfs-tests/tests/functional/rsend/send_realloc_encrypted_files.ksh [hunk 9]
```diff
@@ -15,6 +15,7 @@
 # Copyright (c) 2019 by Lawrence Livermore National Security, LLC.
```
Comment (c0d3z3r0) [inline tests/zfs-tests/tests/functional/rsend/send_realloc_encrypted_files.ksh]: can you fix that, too, when already touching that file, please? :)

File: tests/zfs-tests/tests/functional/rsend/rsend_012_pos.ksh (hunk 8 above)
Comment (Ornias1993) [reply tests/zfs-tests/tests/functional/rsend/rsend_012_pos.ksh]: Space is not the only issue here.

File: tests/zfs-tests/tests/functional/cli_root/zfs_set/property_alias_001_pos.ksh (hunk 7 above)
Comment (Ornias1993) [reply tests/zfs-tests/tests/functional/cli_root/zfs_set/property_alias_001_pos.ksh]: Yeah, Women fashion by Delphix... The name even sounds like a fashion brand for 50+ women.

File: tests/zfs-tests/tests/functional/cli_root/zfs_set/property_alias_001_pos.ksh (hunk 6 above)
Comment (Ornias1993) [reply tests/zfs-tests/tests/functional/cli_root/zfs_set/property_alias_001_pos.ksh]: Don't forget the comma's before inc if we are nitting

File: tests/zfs-tests/tests/functional/compression/compress_004_pos.ksh [hunk 10]
```diff
@@ -21,14 +21,13 @@
 #
 
 #
-# Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
+# Copyright (c) 2007 Sun Microsystems Inc. All rights reserved.
```
Comment (c0d3z3r0) [inline tests/zfs-tests/tests/functional/compression/compress_004_pos.ksh]: 2007,

File: tests/zfs-tests/tests/functional/rsend/send-c_verify_ratio.ksh [hunk 11]
```diff
@@ -13,6 +13,7 @@
 
 #
 # Copyright (c) 2015 by Delphix. All rights reserved.
```
Comment (c0d3z3r0) [inline tests/zfs-tests/tests/functional/rsend/send-c_verify_ratio.ksh]: by

File: tests/zfs-tests/include/properties.shlib [hunk 12]
```diff
@@ -13,8 +13,7 @@
 # Copyright (c) 2012, 2016 by Delphix. All rights reserved.
```
Comment (c0d3z3r0) [inline tests/zfs-tests/include/properties.shlib]: by

File: tests/zfs-tests/include/properties.shlib [hunk 13]
```diff
@@ -10,11 +10,10 @@
 #
 
 #
-# Copyright (c) 2012, 2016 by Delphix. All rights reserved.
+# Copyright (c) 2012, 2016, Delphix. All rights reserved.
 #
 
-typeset -a compress_prop_vals=('on' 'off' 'lzjb' 'gzip' 'gzip-1' 'gzip-2'
-    'gzip-3' 'gzip-4' 'gzip-5' 'gzip-6' 'gzip-7' 'gzip-8' 'gzip-9' 'zle' 'lz4')
+typeset -a compress_prop_vals=('off' 'lzjb' 'lz4' 'gzip' 'zle')
```
Comment (jwk404) [inline tests/zfs-tests/include/properties.shlib]: Are the gzip levels exercised elsewhere, or is this an intentional reduction in coverage?

File: tests/zfs-tests/include/libtest.shlib [hunk 14]
```diff
@@ -20,14 +20,13 @@
...
-# Copyright (c) 2017 Datto Inc.
-# Copyright (c) 2017 Open-E, Inc. All Rights Reserved.
+# Copyright (c) 2009, Sun Microsystems Inc. All rights reserved.
+# Copyright (c) 2012, 2018, Delphix. All rights reserved.
+# Copyright (c) 2017, Tim Chase. All rights reserved.
+# Copyright (c) 2017, Nexenta Systems Inc. All rights reserved.
+# Copyright (c) 2017, Lawrence Livermore National Security LLC. All rights reserved.
+# Copyright (c) 2017, Datto Inc. All rights reserved.
+# Copyright (c) 2017, Open-E Inc. All rights reserved.
 # Use is subject to license terms.
```
Comment (jwk404) [inline tests/zfs-tests/include/libtest.shlib]: Here and throughout, please don't modify existing copyright notices.

Comment (Ornias1993) [reply tests/zfs-tests/include/libtest.shlib]: I didn't change the intention nor the meaning, just formating that was off.
There is no legal consequence.

File: tests/zfs-tests/include/properties.shlib (hunk 13 above)
Comment (Ornias1993) [reply tests/zfs-tests/include/properties.shlib]: The most basic functionality is tested under "history_002_pos.ksh", for all aliasses and levels.
The actual levels should all provide the same coverage.

However, having 9 gzip levels creates 9 times the chance of a gzip level (when randomly selecting) in comparison to any other algorithm.

Considering the future might/will bring other level-based compression, having the levels in here is not sustainable.

File: tests/zfs-tests/include/libtest.shlib (hunk 14 above)
Comment (jwk404) [reply tests/zfs-tests/include/libtest.shlib]: Companies I've worked for have alway advised me not to modify copyrights. I'm not an expert on copyright law, so I won't try to debate whether or not there is legal consequence.

The formatting you've done does change the meaning. By moving the "Use is subject..." line to the bottom, it no longer looks like it applies only to the Sun copyright line. Additionally, it's not uncommon for authors to use editor automation to update or add copyright lines. These changes could inconvenience those authors.

All of which is to say I don't think the pros of the reformat outweigh even the known cons.

Comment (Ornias1993) [reply tests/zfs-tests/include/libtest.shlib]: > Companies I've worked for have alway advised me not to modify copyrights

And I 100% agree with them. Just today I scolded another contributor for changing a licence on a file, without consent of the author.

> By moving the "Use is subject..." line to the bottom, it no longer looks like it applies only to the Sun copyright line.

No it does not:
"Use (of the file) is subject to (the CDDL) licence terms". Has nothing to do with sun or any of the other copyright holders directly. It does not say "is subject to Sun's copyright, or Sun's licence terms".

To be frankfully honest: Legally this is the biggest bullshit ever.... This is a legacy thing from decades ago, when it was required. Use is ALWAYS subject to licence terms and everyone (listed or not) automatically (by law) gains copyright since many years now.

The reason people keep doing it, is simply to keep track of why has rights where (for which this is a great tool!), but legally this is not needed at all.

> Additionally, it's not uncommon for authors to use editor automation to update or add copyright lines

That explains why people screw them up so badly at times. But I need to find the first repo where "my editor automation made me do it" was valued as a valid reason to screw up formating.


TL:DR
These are not "copyrights" these are references. Copyright is governed by (international) law.

Comment (Ornias1993) [reply tests/zfs-tests/include/libtest.shlib]: But okey...
I'll let @behlendorf as chief maintainer for the ZoL repo decide.

File: tests/zfs-tests/tests/functional/compression/compress_003_pos.ksh [hunk 15]
```diff
@@ -62,7 +62,7 @@ typeset -i offset=0
 
 for propname in "compression" "compress"
 do
-	for value in $(get_compress_opts zfs_compress)
+	for value in "${compress_prop_vals[@]:1}"
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/compression/compress_003_pos.ksh]: Is this syntax valid ksh?  I was under the impression `${arr[@]:s:n}` was a bash'ism.  Though I'd be happy to be wrong about that.

File: tests/zfs-tests/tests/functional/compression/compress_004_pos.ksh [hunk 16]
```diff
@@ -94,7 +93,7 @@ typeset -i blknum=0
 
 for propname in "compression" "compress"
 do
-	for value in $(get_compress_opts zfs_compress)
+	for value in "${compress_prop_vals[@]:1}"
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/compression/compress_004_pos.ksh]: Same

File: tests/zfs-tests/include/libtest.shlib (hunk 14 above)
Comment (behlendorf) [reply tests/zfs-tests/include/libtest.shlib]: For the authoritative copyright attribution we've always relied on the git commit history and signed-off-by.

Cosmetically cleaning up these lines isn't an issue.  It does feel a little out of scope for this PR, but aside from that I don't have any objection.

File: tests/zfs-tests/tests/functional/compression/compress_003_pos.ksh (hunk 15 above)
Comment (Ornias1993) [reply tests/zfs-tests/tests/functional/compression/compress_003_pos.ksh]: This actually took me some tries...
If “off” (first in that array) is included it errors out, so I actually had to check to make sure it actually skipped the first value. :)

File: tests/zfs-tests/include/libtest.shlib (hunk 14 above)
Comment (Ornias1993) [reply tests/zfs-tests/include/libtest.shlib]: Yeah I think @c0d3z3r0 was a bit overzealus... reformatting only the actually changed headers would be my prefference also. That being said: its a shame to revert anymore just because “little out of scope” agreed?

File: tests/zfs-tests/tests/functional/compression/compress_003_pos.ksh (hunk 15 above)
Comment (jwk404) [reply tests/zfs-tests/tests/functional/compression/compress_003_pos.ksh]: I had the same thought, and verified that it works on ksh.

Comment (Ornias1993) [reply tests/zfs-tests/tests/functional/compression/compress_003_pos.ksh]: Its nice to know, gives a lot of options!

File: tests/zfs-tests/include/libtest.shlib [hunk 17]
```diff
@@ -20,14 +20,13 @@
...
-# Copyright (c) 2017 by Tim Chase. All rights reserved.
-# Copyright (c) 2017 by Nexenta Systems, Inc. All rights reserved.
-# Copyright (c) 2017 Lawrence Livermore National Security, LLC.
-# Copyright (c) 2017 Datto Inc.
-# Copyright (c) 2017 Open-E, Inc. All Rights Reserved.
+# Copyright (c) 2009, Sun Microsystems Inc. All rights reserved.
+# Copyright (c) 2012, 2018, Delphix. All rights reserved.
+# Copyright (c) 2017, Tim Chase. All rights reserved.
+# Copyright (c) 2017, Nexenta Systems Inc. All rights reserved.
+# Copyright (c) 2017, Lawrence Livermore National Security LLC. All rights reserved.
```
Comment (behlendorf) [inline tests/zfs-tests/include/libtest.shlib]: nit: This line breaks the 79 character limit.  Let's simply remove the duplicate line here and call it a day or drop this hunk entirely.

File: tests/zfs-tests/tests/functional/rsend/send_realloc_encrypted_files.ksh [hunk 18]
```diff
@@ -12,9 +12,11 @@
 #
 
 #
-# Copyright (c) 2019 by Lawrence Livermore National Security, LLC.
+# Copyright (c) 2019, Lawrence Livermore National Security LLC. All rights reserved.
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/rsend/send_realloc_encrypted_files.ksh]: nit: same as above over 79 characters.

File: tests/zfs-tests/include/libtest.shlib (hunk 17 above)
Comment (Ornias1993) [reply tests/zfs-tests/include/libtest.shlib]: I think c style skips commenting...

Comment (Ornias1993) [reply tests/zfs-tests/include/libtest.shlib]: @behlendorf I think you mean "All rights reserved" with "the duplicate line"?

Comment (Ornias1993) [reply tests/zfs-tests/include/libtest.shlib]: Removed "All rights reserved'.
Needs squash later (not at main workstation atm)

File: tests/zfs-tests/tests/functional/rsend/send_realloc_encrypted_files.ksh (hunk 18 above)
Comment (Ornias1993) [reply tests/zfs-tests/tests/functional/rsend/send_realloc_encrypted_files.ksh]: Removed "All rights reserved'.
Needs squash later (not at main workstation atm)

**Comment by Ornias1993** [comment]
@behlendorf 
Squashed and ready for re-review :)
